rosetta/transactions/parse.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package transactions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// ParseTransactions processes the flow transaction and translates it to a list of operations and a list of
+// signers.
+func (p *Parser) ParseTransaction(tx flow.Transaction) ([]object.Operation, []identifier.Account, error) {
+
+	ops := make([]object.Operation, 2)
+
+	if len(tx.Authorizers) == 0 {
+		return nil, nil, fmt.Errorf("invalid authorizers count: %v", len(tx.Authorizers))
+	}
+
+	args := tx.Arguments
+	if len(args) != 2 {
+		return nil, nil, fmt.Errorf("invalid arguments count: %v", len(args))
+	}
+
+	amountArg, err := json.Decode(args[0])
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not parse transaction amount: %w", err)
+	}
+
+	amount := amountArg.String()
+	// Format the amount in the Rosetta format - remove the decimal point and strip
+	// leading zeroes.
+	amount = strings.Replace(amount, ".", "", 1)
+	amount = strings.TrimLeft(amount, "0")
+
+	receiverArg, err := json.Decode(args[1])
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not parse transaction receiver: %w", err)
+	}
+
+	receiver := flow.HexToAddress(receiverArg.String())
+
+	// Create the send operation.
+	ops[0] = object.Operation{
+		ID: identifier.Operation{
+			Index: 0,
+		},
+		RelatedIDs: []identifier.Operation{{Index: 1}},
+		AccountID: identifier.Account{
+			Address: tx.Authorizers[0].String(),
+		},
+		Type:   dps.OperationTransfer,
+		Status: dps.StatusCompleted,
+		Amount: object.Amount{
+			Value: "-" + amount,
+			Currency: identifier.Currency{
+				Symbol:   dps.FlowSymbol,
+				Decimals: dps.FlowDecimals,
+			},
+		},
+	}
+
+	// Create the receive operation.
+	ops[1] = object.Operation{
+		ID: identifier.Operation{
+			Index: 1,
+		},
+		RelatedIDs: []identifier.Operation{{Index: 0}},
+		AccountID: identifier.Account{
+			Address: receiver.String(),
+		},
+		Type:   dps.OperationTransfer,
+		Status: dps.StatusCompleted,
+		Amount: object.Amount{
+			Value: amount,
+			Currency: identifier.Currency{
+				Symbol:   dps.FlowSymbol,
+				Decimals: dps.FlowDecimals,
+			},
+		},
+	}
```

Comment:
WDYT of having a func to build those? It feels weird to have two massive almost identical declarations in a func like that. The only thing that changes in the indexes and that one has a negative amount, IIRC?